zoo_frontend/src/pages/admin/groupDiets/groupDiets.jsx

```diff
@@ -0,0 +1,237 @@
+      return {
+        subenclosures: [],
+        locations: [],
+        error: true,
+        errorMessage: 'Error loading data',
+      };
+    }
+  }
+  static propTypes = {
+    // account: PropTypes.object.isRequired,
+    token: PropTypes.string,
+    classes: PropTypes.object.isRequired,
+    locations: PropTypes.arrayOf(PropTypes.object).isRequired,
+    subenclosures: PropTypes.arrayOf(PropTypes.object).isRequired,
+    error: PropTypes.bool,
+    errorMessage: PropTypes.string,
+  };
+
+  static defaultProps = {
+    token: '',
+    error: false,
+    errorMessage: '',
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      ...generateStateData(props.subenclosures, props.locations),
+    };
+    this.notificationsRef = React.createRef();
+  }
+
+  onRowAdd = (newData) => new Promise(async (resolve, reject) => {
+    const subenclosuresApi = new SubenclosuresAPI(this.props.token);
+    // Reject if a field is not filled out
+    if (!newData.subenclosure || !newData.locationName) {
+      this.notificationsRef.current.showNotification('error', 'Please fill out all of the "Species" and "Scientific Name".');
+      reject();
+      return;
+    }
+
+    // Find the selected location
+    let location = this.props.locations.find((l) => l.location === newData.locationName);
+
+    if (!location) {
+      if (newData.locationName === 'None') {
+        location = { locationId: null };
+      } else {
+        reject();
+        return;
+      }
+    }
+
+    try {
+      // Create the group diet entry
+      await subenclosuresApi.createSubenclosures({
+        subenclosure: newData.subenclosure,
+        locationId: location.locationId,
+      });
+    } catch (err) {
+      reject();
+    }
+
+    // Refresh Data
+    try {
+      const subRes = await subenclosuresApi.getSubenclosures();
+      this.setState({ ...generateStateData(subRes.data, this.props.locations) });
+    } catch (err) {
+      reject();
+      return;
+    }
+    resolve();
+  })
+
+  onRowUpdate = (newData, oldData) => new Promise(async (resolve, reject) => {
+    const subenclosuresApi = new SubenclosuresAPI(this.props.token);
+
+    // Determine if we need to update and what to update
+    let fieldUpdated = false;
+    const updatedFields = {};
+
+    if (newData.subenclosure !== oldData.subenclosure) {
+      fieldUpdated = true;
+      updatedFields.subenclosure = newData.subenclosure;
+    }
+
+    if (newData.locationName !== oldData.locationName) {
+      fieldUpdated = true;
+      let location = this.props.locations.find((l) => l.location === newData.locationName);
+
+      if (!location) {
+        if (newData.locationName === 'None') {
+          location = { locationId: null };
+        } else {
+          reject();
+          return;
+        }
+      }
+
+      updatedFields.locationId = location.locationId;
+    }
+
+    if (fieldUpdated) {
+      await subenclosuresApi.updateSubenclosures(newData.seId, updatedFields);
```

Comment:
Spreading in the state again would knock off an API call does require a map-like iteration to find the updated row, but is likely faster than the API

Reply:
this is how i have been doing it in all the other pages I am just going to leave it for now, it makes sure we have the most up-to-date info all the time